internal/controller/patch_funcs.go

```diff
@@ -169,61 +169,57 @@ func (r *LightrunJavaAgentReconciler) patchAppContainers(lightrunJavaAgent *agen
+	patchedEnvValue := deplAnnotations["lightrun.com/patched-env-value"]
+
+	if patchedEnv != targetEnvVar || patchedEnvValue != agentArg {
+		// If different env was patched before - unpatch it
+		patchedEnvVarIndex := findEnvVarIndex(patchedEnv, container.Env)
+		if patchedEnvVarIndex != -1 {
+			unpatchedEnvValue := unpatchEnvVarValue(container.Env[patchedEnvVarIndex].Value, patchedEnvValue)
+			if unpatchedEnvValue == "" {
+				container.Env = append(container.Env[:patchedEnvVarIndex], container.Env[patchedEnvVarIndex+1:]...)
+			} else {
+				container.Env[patchedEnvVarIndex].Value = unpatchedEnvValue
+			}
 		}
 	}
 
-	if javaToolOptionsIndex == -1 {
+	targetEnvVarIndex := findEnvVarIndex(targetEnvVar, container.Env)
+	if targetEnvVarIndex == -1 {
+		// No such env - add new
 		container.Env = append(container.Env, corev1.EnvVar{
 			Name:  targetEnvVar,
 			Value: agentArg,
 		})
 	} else {
-		if !strings.Contains(container.Env[javaToolOptionsIndex].Value, agentArg) {
-			if len(container.Env[javaToolOptionsIndex].Value+" "+agentArg) > 1024 {
-				return errors.New(targetEnvVar + " has more that 1024 chars")
+		if !strings.Contains(container.Env[targetEnvVarIndex].Value, agentArg) {
+			container.Env[targetEnvVarIndex].Value = container.Env[targetEnvVarIndex].Value + agentArg
+			if len(container.Env[targetEnvVarIndex].Value) > 1024 {
+				return errors.New(targetEnvVar + " has more that 1024 chars. This is a limitation of Java")
 			}
-			container.Env[javaToolOptionsIndex].Value = container.Env[javaToolOptionsIndex].Value + " " + agentArg
 		}
 	}
 	return nil
 }
 
-func (r *LightrunJavaAgentReconciler) unpatchJavaToolEnv(container *corev1.Container, targetEnvVar string, mountPath string, agentCliFlags string) *corev1.Container {
-	agentArg := "-agentpath:" + mountPath + "/agent/lightrun_agent.so"
-	if agentCliFlags != "" {
-		agentArg += "=" + agentCliFlags
+func (r *LightrunJavaAgentReconciler) unpatchJavaToolEnv(deplAnnotations map[string]string, container *corev1.Container) *corev1.Container {
+	patchedEnv := deplAnnotations["lightrun.com/patched-env-name"]
+	patchedEnvValue := deplAnnotations["lightrun.com/patched-env-value"]
+	if patchedEnv == "" && patchedEnvValue == "" {
+		return container
 	}
 
-	var updatedSlice []corev1.EnvVar
-	for _, envVar := range container.Env {
-		if envVar.Name == targetEnvVar {
-			var updatedEnv []string
-			optArray := strings.Split(envVar.Value, " ")
-			for _, opt := range optArray {
-				if opt != agentArg {
-					updatedEnv = append(updatedEnv, opt)
-				}
-			}
-			if len(updatedEnv) > 0 {
-				envVar.Value = strings.Join(updatedEnv, " ")
-				updatedSlice = append(updatedSlice, envVar)
-			}
+	envVarIndex := findEnvVarIndex(patchedEnv, container.Env)
+	if envVarIndex != -1 {
+		value := strings.ReplaceAll(container.Env[envVarIndex].Value, patchedEnvValue, "")
+		if value == "" {
+			container.Env = append(container.Env[:envVarIndex], container.Env[envVarIndex+1:]...)
```

Comment:
Well this confused me a bit but with some help of non-human :) I managed to get it , the logic is basically to remove environment variable from container environment variable list by applying technics of slicing all elements before but not include the index of the removed env var and  all elements from the element immediately after the one at index removed environment variable to the end.

maybe we can add comment or extract this logic to function name removeEnvVar

Reply:
actually, there is `slices.Delete` was added in go 1.21 
we can use it 👍